sdk/appcenter/src/main/java/com/microsoft/appcenter/http/DefaultHttpClientCallTask.java

```diff
@@ -0,0 +1,337 @@
+                }
+
+                /* Send payload on the wire. */
+                urlConnection.setDoOutput(true);
+                urlConnection.setFixedLengthStreamingMode(binaryPayload.length);
+                OutputStream out = urlConnection.getOutputStream();
+
+                //noinspection TryFinallyCanBeTryWithResources
+                try {
+                    writePayload(out, binaryPayload);
+                } finally {
+                    out.close();
+                }
+            }
+            if (isCancelled()) {
+                return null;
+            }
+
+            /* Read response. */
+            int status = urlConnection.getResponseCode();
+            String response = readResponse(urlConnection);
+            if (AppCenterLog.getLogLevel() <= Log.VERBOSE) {
+                String contentType = urlConnection.getHeaderField(CONTENT_TYPE_KEY);
+                String logPayload;
+                if (contentType == null || contentType.startsWith("text/") || contentType.startsWith("application/")) {
+                    logPayload = TOKEN_REGEX_JSON.matcher(response).replaceAll("token\":\"***\"");
+                } else {
+                    logPayload = "<binary>";
+                }
+                AppCenterLog.verbose(LOG_TAG, "HTTP response status=" + status + " payload=" + logPayload);
+            }
+
+            /* Accept all 2xx codes. */
+            if (status >= 200 && status < 300) {
+                return response;
+            }
+
+            /* Generate exception on failure. */
+            throw new HttpException(status, response);
+        } finally {
+
+            /* Release connection. */
+            urlConnection.disconnect();
+        }
+    }
+
+    private boolean isHttpResult(Object result) {
+        return result instanceof String || result instanceof HttpException;
+    }
+
+    @Override
+    protected Object doInBackground(Void... params) {
+
+        /* Do tag socket to avoid strict mode issue. */
+        TrafficStats.setThreadStatsTag(THREAD_STATS_TAG);
+        try {
+            return doHttpCall();
+        } catch (Exception e) {
+            return e;
+        } finally {
+            TrafficStats.clearThreadStatsTag();
+        }
+    }
+
+    @Override
+    protected void onPreExecute() {
+        mTracker.onStart(this);
+    }
+
+    @Override
+    protected void onPostExecute(Object result) {
+        mTracker.onFinish(this);
+        if (result instanceof Exception) {
+            mServiceCallback.onCallFailed((Exception) result);
+        } else {
+            mServiceCallback.onCallSucceeded(result.toString());
+        }
+    }
+
+    @Override
+    protected void onCancelled(Object result) {
+
+        /* Handle the result even if it was cancelled. */
+        if (isHttpResult(result)) {
+            onPostExecute(result);
+        } else {
+            mTracker.onFinish(this);
+        }
+    }
+
+    interface Tracker {
```

Comment:
This definition seems weird. Should it be defined in DefaultHttpClient.java?

Reply:
I don't think so. It's callback interface for this class.